File: tools/hack_wavecrypt.cpp
#include "wavecrypt/grid/grid_2d.hpp"
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <random>
#include <iomanip>
#include <chrono>

using namespace wavecrypt::grid;

// Same rudimental absorption as wavehash_cli
std::vector<uint8_t> wavehash(const std::string& message) {
    Grid2D grid(16, 16);
    if (!message.empty()) {
        std::size_t idx = 0;
        for (std::size_t y = 0; y < grid.height(); ++y) {
            for (std::size_t x = 0; x < grid.width(); ++x) {
                uint32_t val = static_cast<uint32_t>(message[idx % message.length()]);
                val ^= (val << 8) | (static_cast<uint32_t>(idx) * 0x9E3779B9); 
                grid.inject_energy(x, y, val);
                idx++;
            }
        }
    }
    for (int i = 0; i < 25; ++i) grid.propagate();
    return grid.collapse_to_bytes(32);
}

std::string bytes_to_hex(const std::vector<uint8_t>& bytes) {
    std::string hex;
    const char hex_chars[] = "0123456789abcdef";
    for(auto b : bytes) {
        hex.push_back(hex_chars[b >> 4]);
        hex.push_back(hex_chars[b & 0x0F]);
    }
    return hex;
}

std::string random_string(std::size_t length) {
    static const char alphanum[] = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
    std::string s;
    for (std::size_t i = 0; i < length; ++i) {
        s += alphanum[static_cast<std::size_t>(rand()) % (sizeof(alphanum) - 1)];
    }
    return s;
}

void test_avalanche() {
    std::cout << "\n--- Avalanche Effect Test ---\n";
    std::cout << "Goal: Changing 1 bit should change ~50% (128) of the hash bits.\n";
    
    std::string msg1 = "Hello, WaveCrypt! This is a test.";
    std::string msg2 = "Iello, WaveCrypt! This is a test."; // Changed H to I
    
    auto h1 = wavehash(msg1);
    auto h2 = wavehash(msg2);
    
    std::cout << "Base Msg : '" << msg1 << "'\n";
    std::cout << "Hash 1   : " << bytes_to_hex(h1).substr(0, 32) << "...\n";
    std::cout << "Mod Msg  : '" << msg2 << "'\n";
    std::cout << "Hash 2   : " << bytes_to_hex(h2).substr(0, 32) << "...\n";
    
    int diff_count = 0;
    for (std::size_t i = 0; i < h1.size(); ++i) {
        uint8_t xor_val = h1[i] ^ h2[i];
        // Count set bits
        while (xor_val) {
            diff_count += xor_val & 1;
            xor_val >>= 1;
        }
    }
    
    double percentage = (static_cast<double>(diff_count) / 256.0) * 100.0;
    std::cout << "\nBits flipped: " << diff_count << " out of 256 (" << percentage << "%)\n";
    
    if (percentage >= 45.0 && percentage <= 55.0) {
        std::cout << "Result: PASS (Excellent Avalanche)\n";
    } else {
        std::cout << "Result: FAIL (Avalanche is biased)\n";
    }
}

void test_collision() {
    std::cout << "\n--- Collision (Birthday) Attack ---\n";
    std::cout << "Goal: Find two strings with SAME hash.\n";
    std::cout << "Testing 1,000,000 random strings in C++...\n";
    
    std::unordered_map<std::string, std::string> seen;
    int collisions = 0;
    
    srand(42);
    
    for (int i = 0; i < 100000; ++i) {
        if (i % 25000 == 0 && i > 0) std::cout << "  ... tried " << i << " hashes\n";
        std::string r_str = random_string(8);
        auto h_bytes = wavehash(r_str);
        std::string h_str((char*)h_bytes.data(), h_bytes.size());
        
        if (seen.count(h_str) && seen[h_str] != r_str) {
            std::cout << "\nCOLLISION FOUND!\n";
            std::cout << "String 1: " << seen[h_str] << "\n";
            std::cout << "String 2: " << r_str << "\n";
            collisions++;
            break;
        }
        seen[h_str] = r_str;
    }
    
    if (collisions == 0) {
        std::cout << "Result: PASS (No full collisions found in 1 Million attempts)\n";
    } else {
        std::cout << "Result: FAIL (Hash collision detected!)\n";
    }
}

int main() {
    std::cout << "Initiating C++ Hack Sequence...\n";
    test_avalanche();
    test_collision();
    return 0;
}

File: scripts/hack_wavecrypt.py
import os
import subprocess
import sys
import random
import string
from collections import defaultdict

CLI_PATH = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', 'build', 'tools', 'wavehash_cli'))

def wavehash(message):
    result = subprocess.run([CLI_PATH, message], capture_output=True, text=True)
    if result.returncode != 0:
        print("Hashing failed!")
        sys.exit(1)
    return result.stdout.strip()

def hex_to_bin(hex_str):
    """Convert hex string to a binary string of exactly 256 bits."""
    return bin(int(hex_str, 16))[2:].zfill(256)

def test_avalanche():
    print("\n--- Avalanche Effect Test ---")
    print("Goal: Changing 1 character should change ~50% (128) of the bits.")
    
    base_msg = "Hello, WaveCrypt! This is a test."
    hash1_hex = wavehash(base_msg)
    hash1_bin = hex_to_bin(hash1_hex)
    
    # Change exactly one bit (the first character 'H' -> 'I')
    mod_msg = "Iello, WaveCrypt! This is a test."
    hash2_hex = wavehash(mod_msg)
    hash2_bin = hex_to_bin(hash2_hex)
    
    print(f"Base Msg : '{base_msg}'")
    print(f"Hash 1   : {hash1_hex[:32]}...")
    
    print(f"Mod Msg  : '{mod_msg}'")
    print(f"Hash 2   : {hash2_hex[:32]}...")
    
    # Count bit differences
    diff_count = sum(1 for a, b in zip(hash1_bin, hash2_bin) if a != b)
    percentage = (diff_count / 256.0) * 100
    
    print(f"\nBits flipped: {diff_count} out of 256 ({percentage:.2f}%)")
    if 45.0 <= percentage <= 55.0:
        print("Result: PASS (Excellent Avalanche)")
    else:
        print("Result: FAIL (Avalanche is biased)")

def test_collision():
    print("\n--- Collision (Birthday) Attack ---")
    print("Goal: Find two different strings that produce the SAME hash.")
    print("Testing 2,500 random strings...")
    
    seen = {}
    collisions = 0
    
    # Generate random 8-char strings
    for i in range(2500):
        if i % 500 == 0 and i > 0:
            print(f"  ... tried {i} hashes")
            
        r_str = ''.join(random.choices(string.ascii_letters + string.digits, k=8))
        h = wavehash(r_str)
        
        # We check full 256-bit collision
        if h in seen and seen[h] != r_str:
            print(f"\nCOLLISION FOUND!")
            print(f"String 1: {seen[h]}")
            print(f"String 2: {r_str}")
            print(f"Hash: {h}")
            collisions += 1
            break
            
        seen[h] = r_str
        
    if collisions == 0:
        print("Result: PASS (No full collisions found in 100k attempts)")
    else:
        print("Result: FAIL (Hash collision detected!)")
        
    # Extra: Truncated collision (testing diffusion weakness)
    print("\nTesting 16-bit truncated collisions (Checking for structural weakness)...")
    trunc_seen = defaultdict(list)
    for k, v in seen.items():
        trunc = k[:4] # First 16 bits (4 hex chars)
        trunc_seen[trunc].append(v)
        
    many_collisions = sum(1 for k, v in trunc_seen.items() if len(v) > 1)
    print(f"Found {many_collisions} truncated collisions. (This is expected in short lengths).")

if __name__ == "__main__":
    if not os.path.exists(CLI_PATH):
        print(f"Error: Could not find {CLI_PATH}. Did you compile it?")
        sys.exit(1)
        
    print("Initiating Hack Sequence...")
    test_avalanche()
    test_collision()
